onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,401 @@
+     M               A  B  C  [Inputs]
+     |                \ |  /
+    Cast               Gemm
+     |                / |   \
+     |               /  |    \
+     |              /   |     \
+     |          Slice  Slice  Slice
+  Identity        |     |       |
+     |             -----        |
+     -----------    |           |
+                 \  |           |
+                  Gemm          |
+                    |           |
+                    |           |
+                Softmax         |
+                    |          /
+                    |         /
+                     \       /
+                       \    /
+                        Gemm
+                          |
+                   ActivationLinear
+                          |
+                        Output [Final ouput]
+ This kernel creates a DML_GRAPH, as mentioned above.
+ For refernce refer to this Doc: 
+ https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.Attention
+ */
+namespace Dml
+{
+class DmlOperatorAttention : public DmlOperator
+{
+public:
+    DmlOperatorAttention(const MLOperatorKernelCreationContext& kernelCreationContext)
+    :   DmlOperator(kernelCreationContext)
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() >= 3);
+        DmlOperator::Initialize(kernelCreationContext, std::nullopt, std::nullopt, std::nullopt, std::nullopt, 1);
+
+        std::vector<uint32_t> inputTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(0);
+        std::vector<uint32_t> weightTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(1);
+        std::vector<uint32_t> biasTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(2);
+        std::vector<uint32_t> maskIndexTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(3);
+        ML_CHECK_VALID_ARGUMENT(inputTensorShape.size() == 3);
+        ML_CHECK_VALID_ARGUMENT(weightTensorShape.size() == 2);
+        ML_CHECK_VALID_ARGUMENT(biasTensorShape.size() == 1);
+        ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] == biasTensorShape[0]);
+        ML_CHECK_VALID_ARGUMENT(biasTensorShape[0] % 3 == 0);
+        ML_CHECK_VALID_ARGUMENT(inputTensorShape[2] == weightTensorShape[0]);
+        ML_CHECK_VALID_ARGUMENT(maskIndexTensorShape.size() > 1); // TODO: fix Attention kernel when maskIndexTensorShape is 1
+        const uint32_t batchSize = inputTensorShape[0];
+        const uint32_t sequenceLength = inputTensorShape[1];
+        const uint32_t hiddenSize = biasTensorShape[0] / 3;
+        const uint32_t numHeads = gsl::narrow_cast<uint32_t>(kernelCreationContext.GetAttribute<int64_t>(AttrName::NumHeads));
+        ML_CHECK_VALID_ARGUMENT(hiddenSize % numHeads == 0);
+        const uint32_t headSize = hiddenSize / numHeads;
+
+        uint32_t desiredWeightTensorShape[3] {batchSize, weightTensorShape[0], 3 * hiddenSize};
```

Comment:
>3 * hiddenSize [](http://example.com/codeflow?start=79&length=14)

[minor] Rather than recompute `hiddenSize * 3`, since hiddenSize is just size / 3 in the first place, assign `uint32_t biasSize = biasTensorShape[0]` and use that here?

Reply:
I think w.r.t. Attention kernel logic, writing it explicitly as 3*hiddenSize makes it clearer that the given Weight tensor contains weight for 3 tensors (Query, Key and Value). 
This understanding won't be clear very much if directly use biasTensorShape[0].
Thoughts?

Reply:
Shrug.